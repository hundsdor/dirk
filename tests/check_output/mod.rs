use std::{path::PathBuf, process::Command};

use insta::{assert_snapshot, with_settings, Settings};
use tempdir::TempDir;

pub(crate) fn test_main(path: &str, name: &str) {
    let mut cmd = Command::new(env!("CARGO"));

    cmd.arg("run");
    cmd.arg("--example");
    cmd.arg(name);

    cmd.env("RUSTFLAGS", "--cap-lints=allow");
    cmd.env("CARGO_TERM_COLOR", "never");

    let mut current_dir = PathBuf::new();
    current_dir.push("test-data");
    current_dir.push(path);

    cmd.current_dir(&current_dir);

    let target_dir = TempDir::new_in(env!("CARGO_TARGET_TMPDIR"), name).unwrap();
    cmd.env("TARGET_DIR", target_dir.path());

    let output = cmd
        .output()
        .unwrap_or_else(|_| panic!("Failed to execute test {name}"));

    let mut settings = Settings::clone_current();

    let filter_block_package_cache = r".*Blocking waiting for file lock on package cache.*\n";
    settings.add_filter(filter_block_package_cache, "");

    let filter_block_build_directory = r".*Blocking waiting for file lock on build directory.*\n";
    settings.add_filter(filter_block_build_directory, "");

    let filter_finished_targets = r".*Finished .* target\(s\) in .*s";
    settings.add_filter(filter_finished_targets, "Finished compiling target(s)");

    let filter_compiling = r".*Compiling .*\n";
    settings.add_filter(filter_compiling, "");

    let filter_updating_index = r".*Updating crates.io index\n";
    settings.add_filter(filter_updating_index, "");

    let filter_locking = r".*Locking \d+ packages to latest compatible versions\n";
    settings.add_filter(filter_locking, "");

    let filter_adding = r".*Adding .* (\(.*\))?\n";
    settings.add_filter(filter_adding, "");

    let filter_path = if cfg!(windows) {
        r"(?:([\w\-_]+)(?:\\))+([\w\-_]+?\.[\w\-_]+)?"
    } else {
        r"(?:([\w\-_]+)(?:/))+([\w\-_]+?\.[\w\-_]+)?"
    };
    settings.add_filter(filter_path, "$1/$2");

    if cfg!(windows) {
        // `\\` has already been replaced by `/`
        let filter_exe = r"Running `(?:([\w\-_]+)(?:/))+([\w\-_]+?).exe`";
        settings.add_filter(filter_exe, "Running `$1/$2`");
    }

    let pretty = format!(
        "Stdout:\n{}\n\nStderr:\n{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    );

    settings.bind(|| {
        with_settings!({snapshot_suffix => name}, {
            assert_snapshot!(pretty);
        });
    });
}
