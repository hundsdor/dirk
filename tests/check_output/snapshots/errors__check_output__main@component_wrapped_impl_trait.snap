---
source: tests/check_output/mod.rs
expression: pretty
---
Stdout:


Stderr:
error[E0666]: nested `impl Trait` is not allowed
  --> examples/component_wrapped_impl_trait.rs:19:50
   |
17 | / #[component(
18 | |     [
19 | |         cookies: cloned_instance_bind(Rc<RefCell<impl Cookies>>),
   | |                                                  ^^^^^^^^^^^^ nested `impl Trait` here
20 | |         user_name: cloned_instance_bind(U),
21 | |         application: static_bind(Application<C, U>) [cookies, user_name]
22 | |     ]
23 | | )]
   | |__- outer `impl Trait`

error[E0562]: `impl Trait` is not allowed in impl headers
  --> examples/component_wrapped_impl_trait.rs:19:50
   |
19 |         cookies: cloned_instance_bind(Rc<RefCell<impl Cookies>>),
   |                                                  ^^^^^^^^^^^^
   |
   = note: `impl Trait` is only allowed in arguments and return types of functions and methods

error[E0562]: `impl Trait` is not allowed in field types
  --> examples/component_wrapped_impl_trait.rs:19:50
   |
19 |         cookies: cloned_instance_bind(Rc<RefCell<impl Cookies>>),
   |                                                  ^^^^^^^^^^^^
   |
   = note: `impl Trait` is only allowed in arguments and return types of functions and methods

error[E0277]: the trait bound `StaticFactoryApplication<impl Cookies, U>: Provider<Application<C, U>>` is not satisfied
  --> examples/component_wrapped_impl_trait.rs:21:9
   |
21 |         application: static_bind(Application<C, U>) [cookies, user_name]
   |         ^^^^^^^^^^^ the trait `Provider<Application<C, U>>` is not implemented for `StaticFactoryApplication<impl Cookies, U>`
   |
   = note: required for the cast from `Rc<StaticFactoryApplication<impl Cookies, U>>` to `Rc<(dyn Provider<Application<C, U>> + 'static)>`

Some errors have detailed explanations: E0277, E0562, E0666.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `application` (example "component_wrapped_impl_trait") due to 4 previous errors
